### PR TITLE
support deploy azure environment, secret, variables and refactoring

### PR DESCRIPTION
**Pull request by squirrelsc**
1. support azure deployment and add arm template. It deploys only, not
   retrieve remote information. And support limited configurations.
2. support variables from runbook, separated file, command line and env.
3. support secret, swap them from log.
4. remove envs, use constants to pass variables like path.
5. add clone to copy from runbook to actual environment.
6. refine runner to delete env on failure.
7. refine schema to support listable validator, and rich configs.
8. change most config, schema, data to runbook for clearer term
9. other minor improvements.

### CODE REVIEW

File: lisa/commands.py [hunk 1]
```diff
@@ -2,26 +2,31 @@
...
+    replace_variables,
+)
 
 _get_init_logger = functools.partial(get_logger, "init")
 
 
-def _load_extends(base_path: Path, extends_config: Dict[str, object]) -> None:
-    for p in cast(List[str], extends_config.get(constants.PATHS, list())):
+def _load_extends(base_path: Path, extends_runbook: schema.Extension) -> None:
+    for p in extends_runbook.paths:
```
Comment (andyleejordan) [inline lisa/commands.py:29]: 👍

File: lisa/parameter_parser/runbook.py [hunk 2]
```diff
@@ -13,35 +14,39 @@
 
 _schema: Optional[Schema] = None
 
+_get_init_logger = partial(get_logger, "init", "runbook")
```
Comment (andyleejordan) [inline lisa/parameter_parser/runbook.py:17]: Love it.

File: lisa/secret.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    if origin and origin not in _secret_set:
+        if not isinstance(origin, str):
+            origin = str(origin)
+        _secret_set.add(origin)
+        _secret_list.append((origin, replace(origin, sub=sub, mask=mask)))
+        # deal with longer first, in case it's broken by shorter
+        _secret_list = sorted(_secret_list, reverse=True, key=lambda x: len(x[0]))
+
+
+def mask(input: str) -> str:
```
Comment (andyleejordan) [inline lisa/secret.py:60]: We're going to start to want to include unit tests (or even just embedded tests via doctest). This _looks_ like it works to me.

Comment (andyleejordan) [reply lisa/secret.py:60]: Nevermind, I found the tests further down 😅

File: lisa/tests/test_secret.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        add_secret("t1t2", sub="**")
+        result = mask("t1t2 t1 test3")
+        self.assertEqual(result, "** * test3")
+
+    def test_default_mask(self) -> None:
+        add_secret("test1")
+        result = mask("test1 test3")
+        self.assertEqual(result, "****** test3")
+
+    def test_partten(self) -> None:
```
Comment (andyleejordan) [inline lisa/tests/test_secret.py]: typo, 'partten'

File: lisa/tests/test_secret.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        add_secret("test", sub="*")
+        add_secret(
+            "f5132846-2aff-4726-baea-8c480ce9eb06",
+            mask=re.compile(
+                r"^([0-9a-f]{8})-(?:[0-9a-f]{4}-){3}[0-9a-f]{8}([0-9a-f]{4})$"
+            ),
+            sub=r"\1-****-****-****-********\2",
+        )
+        result = mask("my test f5132846-2aff-4726-baea-8c480ce9eb06 not")
+        self.assertEqual(result, "my * f5132846-****-****-****-********eb06 not")
```
Comment (andyleejordan) [inline lisa/tests/test_secret.py:38]: Ah, nifty. Can reinclude matched groups in the output for partial matching. I see.

File: lisa/tests/test_secret.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    def test_built_in_pattern(self) -> None:
+        add_secret("test", sub="*")
+        add_secret("f5132846-2aff-4726-baea-8c480ce9eb06", mask=PATTERN_GUID)
+        result = mask("my test f5132846-2aff-4726-baea-8c480ce9eb06 not")
+        self.assertEqual(result, "my * f5132846-****-****-****-********eb06 not")
+
+    def test_fallback(self) -> None:
+        add_secret(
+            "test1", mask=re.compile(r"^doesn'tmatch$"), sub=r"*****",
```
Comment (andyleejordan) [inline lisa/tests/test_secret.py]: This helped make the earlier test make more sense.

File: lisa/tests/test_variables.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+from lisa.util.logger import get_logger
+
+
+class VariableTestCase(TestCase):
+    def setUp(self) -> None:
+        secret.reset()
+
+    def test_in_env(self) -> None:
+        os.environ["LISA_normalValue"] = "value_from_env"
+        os.environ["S_LISA_normalEntry"] = "s_value_from_env"
```
Comment (andyleejordan) [inline lisa/tests/test_variables.py:17]: How much configuration are we really expecting to set via environment variables? LISAv2 takes almost no input in this manner, instead preferring explicitly set variables on the command line.

**Review by andyleejordan — APPROVED** [review]
Looks good!

File: lisa/tests/test_secret.py (hunk 4 above)
Comment (squirrelsc) [reply lisa/tests/test_secret.py]: fixed

File: lisa/tests/test_variables.py (hunk 7 above)
Comment (squirrelsc) [reply lisa/tests/test_variables.py:17]: I feel env variables are more secure than command line. The command line are copy/paste, and easy to disclose. Github Action and ADO supports to pass secret values through env variables. That keeps env variables not be touched by command line users.

**Pull request merged**